src/main/kotlin/io/perfometer/dsl/HttpDsl.kt

```diff
@@ -1,117 +1,130 @@
     }
 
-    private fun request(httpMethod: HttpMethod, urlString: String?): RequestBuilder {
+    private fun request(httpMethod: HttpMethod, urlString: String?, builder: RequestDsl.() -> Unit) {
         val requestUrl = urlString?.toUrl() ?: baseURL
-        val request = RequestBuilder(requestUrl, httpMethod, headers)
-        steps.add(RequestStep(request))
-        return request
+        val request = RequestDsl(requestUrl, httpMethod, headers).apply(builder).build()
+        scenarioRunner.runStep(RequestStep(request))
     }
 
-    fun get(urlString: String? = null) = request(HttpMethod.GET, urlString)
-    fun post(urlString: String? = null) = request(HttpMethod.POST, urlString)
-    fun put(urlString: String? = null) = request(HttpMethod.PUT, urlString)
-    fun delete(urlString: String? = null) = request(HttpMethod.DELETE, urlString)
-    fun patch(urlString: String? = null) = request(HttpMethod.PATCH, urlString)
+    fun get(urlString: String? = null, builder: RequestDsl.() -> Unit)
+            = request(HttpMethod.GET, urlString, builder)
+    fun post(urlString: String? = null, builder: RequestDsl.() -> Unit)
+            = request(HttpMethod.POST, urlString, builder)
+    fun put(urlString: String? = null, builder: RequestDsl.() -> Unit)
+            = request(HttpMethod.PUT, urlString, builder)
+    fun delete(urlString: String? = null, builder: RequestDsl.() -> Unit)
+            = request(HttpMethod.DELETE, urlString, builder)
+    fun patch(urlString: String? = null, builder: RequestDsl.() -> Unit)
+            = request(HttpMethod.PATCH, urlString, builder)
 
     fun basicAuth(user: String, password: String) {
         val credentialsEncoded = Base64.getEncoder().encodeToString("$user:$password".toByteArray())
-        header { HttpHeaders.AUTHORIZATION to "Basic $credentialsEncoded" }
+        header(HttpHeaders.AUTHORIZATION to "Basic $credentialsEncoded")
     }
 
     fun pause(duration: Duration) {
-        steps.add(PauseStep(duration))
+        scenarioRunner.runStep(PauseStep(duration))
     }
 }
 
-class ScenarioBuilder(
+class Scenario(
         private val baseURL: URL,
         private val builder: HttpDsl.() -> Unit,
 ) {
-    fun build(): Scenario {
-        return Scenario(HttpDsl(baseURL).apply(builder).scenarioSteps)
+    private var runner: ScenarioRunner = CoroutinesScenarioRunner(SimpleHttpClient(false))
+
+    fun runner(runner: ScenarioRunner): Scenario {
+        this.runner = runner
+        return this
+    }
+
+    fun run(userCount: Int): ScenarioStatistics {
+        runner.runUsers(userCount) {
+            HttpDsl(baseURL, runner).apply(builder)
+        }
+        return runner.statistics()
     }
 }
 
 fun scenario(baseUrlString: String,
-             builder: HttpDsl.() -> Unit): ScenarioBuilder {
-    val baseUrl: URL = baseUrlString.toUrl()
-    return ScenarioBuilder(baseUrl, builder)
-}
+             builder: HttpDsl.() -> Unit): Scenario = Scenario(baseUrlString.toUrl(), builder)
```

Comment:
Cosmetics: As above, the type is clear from the constructor call, no need to declare it in the function signature.

Reply:
done